web/webkit/src/test/scala/net/liftweb/http/FutureWithSessionSpec.scala

```diff
@@ -0,0 +1,290 @@
+
+  "FutureWithSession" should {
+
+    "fail if session is not available" in {
+      val future = FutureWithSession.withCurrentSession("kaboom")
+
+      future.value must eventually(beSome(beFailedTry[String].withThrowable[IllegalStateException](
+        "LiftSession not available in this thread context"
+      )))
+    }
+
+    "succeed with original value if session is available" withSFor "/" in {
+      val future = FutureWithSession.withCurrentSession("works!")
+
+      future.value must eventually(beEqualTo(Some(Success("works!"))))
+    }
+
+    "have access to session variables in Future task" withSFor "/" in {
+      SessionVar1("dzien dobry")
+
+      val future = FutureWithSession.withCurrentSession(SessionVar1.is)
+
+      future.value must eventually(beEqualTo(Some(Success("dzien dobry"))))
+    }
+
+    "have access to request variables in Future task" withSFor "/" in {
+      ReqVar1("guten tag")
+
+      val future = FutureWithSession.withCurrentSession(ReqVar1.is)
+
+      future.value must eventually(beEqualTo(Some(Success("guten tag"))))
+    }
+
+    "have access to session variables in onComplete()" withSFor "/" in {
+      // workaround for a possible race condition in SessionVar
+      // https://groups.google.com/forum/#!topic/liftweb/V1pWy14Wl3A
+      SessionVar1.is
+
+      val future = FutureWithSession.withCurrentSession("thorgal")
+      future.onComplete {
+        case Success(v) => SessionVar1(v)
+        case Failure(reason) => ko("Future execution failed: " + reason)
+      }
+
+      SessionVar1.is must eventually(beEqualTo("thorgal"))
+    }
+
+    "have access to request variables in onComplete()" withSFor "/" in {
+      val future = FutureWithSession.withCurrentSession("thor")
+      future.onComplete {
+        case Success(v) => ReqVar1(v)
+        case Failure(reason) => ko("Future execution failed: " + reason)
+      }
+
+      ReqVar1.is must eventually(beEqualTo("thor"))
+    }
+
+    "have access to session variables in chains of map()" withSFor "/" in {
+      SessionVar1("b")
+      SessionVar2("c")
+
+      val future = FutureWithSession.withCurrentSession("a")
+      val mapped = future.map(_ + SessionVar1.is).map(_ + SessionVar2.is)
+
+      mapped.value must eventually(beEqualTo(Some(Success("abc"))))
+    }
+
+    "have access to request variables in chains of map()" withSFor "/" in {
+      ReqVar1("b")
+      ReqVar2("c")
+
+      val future = FutureWithSession.withCurrentSession("a")
+      val mapped = future.map(_ + ReqVar1.is).map(_ + ReqVar2.is)
+
+      mapped.value must eventually(beEqualTo(Some(Success("abc"))))
+    }
+
+    "have access to session variables in chains of flatMap()" withSFor "/" in {
+      SessionVar1("e")
+      SessionVar2("f")
+
+      val future = FutureWithSession.withCurrentSession("d")
+      val mapped = future
+        .flatMap { s => val out = s + SessionVar1.is; Future(out) }
+        .flatMap { s => val out = s + SessionVar2.is; Future(out) }
```

Comment:
It's just that it's usually a code smell. 😬 

Also, isn't this basically the same as a for-comp?